torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -449,6 +458,53 @@ void ProcessGroupAgent::listenLoop() {
   }
 }
 
+void ProcessGroupAgent::pollTimedOutRPCs() {
+  while (!shutdown_.load()) {
+    std::this_thread::sleep_for(std::chrono::seconds(5));
```

Comment:
You can determine the time to wait by looking at the first entry in the ordered futures map. Then you either wait that long, or until you're woken up by a condition variable if there are no futures to begin with.

Reply:
@pietern, for this suggestion, should we add a new condition variable that wakes up this thread whenever there is a new future created? There is currently a `futureCV` but this notifies when a future has been received and marked as completed.

Reply:
Yes, that sounds good. In order to guarantee correctness we have to do this. There are all sorts of edge cases that we'd fail to cover otherwise. We only need to wait on this one if there are no futures to begin with, though. If there are, we can use the timeout of the first one. This assumes that it never happens that you first call fut1 with timeout=5 and right after that call fut2 with timeout=1 -- it assumes the absolute timestamp of timeouts will only increase. If we were to allow for this case, the logic becomes more complex and contentious, because you'd have to wake up every time a future is created, which seems wasteful to me.